### PR TITLE
Introduce IEx.Server.run/1

### LINKED ISSUE #8203: Should IEx.Server be considered a private API?
While building a feature for NervesHub, I've wrapped IEx.Server with a new `group_leader` to allow for tunneling IO over a Phoenix Channel. This all _feels_ alright, but I know with a recent Elixir version upgrade, some things that were considered private changed breaking some projects. I'd like to prevent this (even though there is only _one_ call into `IEx.Server`)

For what it's worth [here](https://github.com/nerves-hub/nerves_hub/blob/fa79778d32830821c6077cc759340ceeee6748b4/lib/nerves_hub/console_channel.ex) is the implementation.

### PR DESCRIPTION
**Pull request by josevalim**
Since now there may be multiple servers running,
we also introduced IEx.Broker to negociate pry
requests across multiple sessions.

Shell sessions though are tracked separately from
the remaining ones, as we delegate the shell control
to Erlang's user_drv.

Closes #8203.

### CODE REVIEW

File: lib/iex/lib/iex/broker.ex [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  end
+
+  def respond(broker_pid, take_ref, false) do
+    GenServer.call(broker_pid, {:refuse, take_ref})
+  end
+
+  @doc """
+  Client requests a take over.
+  """
+  @spec take_over(binary, keyword) :: {:ok, pid, pid} | {:error, :no_iex | :refused}
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]: `{:ok, pid, pid}` -> `{:ok, server_pid :: pid, leader_pid :: pid}`?

File: lib/iex/lib/iex/broker.ex [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  @doc """
+  Finds the IEx server running inside `:user_drv`, on this node exclusively.
+  """
+  @spec shell :: pid | nil
+  def shell() do
+    # Locate top group leader, always registered as user
+    # can be implemented by group (normally) or user
+    # (if oldshell or noshell)
+    if user = Process.whereis(:user) do
```
Comment (jeregrine) [inline lib/iex/lib/iex/broker.ex]: do we need an if here? Could we just match on the user and fail?

File: lib/iex/test/iex/server_test.exs [hunk 3]
```diff
@@ -3,65 +3,185 @@ Code.require_file("../test_helper.exs", __DIR__)
...
+        receive do: (:run -> :ok)
+        #{session}
+        """)
+      end)
+
+    assert_receive {:running, evaluator}
+    {task, evaluator}
+  end
+
+  defp pry_request(how_many) do
```
Comment (jeregrine) [inline lib/iex/test/iex/server_test.exs]: this is a pretty beautiful solution

File: lib/iex/lib/iex/broker.ex [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+  def handle_call({:take_over, identifier, opts}, {_, ref} = from, state) do
+    case servers(state) do
+      [] ->
+        {:reply, {:error, :no_iex}, state}
+
+      servers ->
+        server_refs =
+          for {server_ref, server_pid} <- servers do
+            send(server_pid, {:take_over, self(), {ref, server_ref}, identifier, opts})
```
[the remark this answers is not in the document]
Comment (fertapric) [reply lib/iex/lib/iex/broker.ex]: The documentation of `respond` says `Note we need to pass the broker pid to support remote shells.`

File: lib/iex/lib/iex/broker.ex [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  Client requests a take over.
+  """
+  @spec take_over(binary, keyword) ::
+          {:ok, server :: pid, group_leader :: pid} | {:error, :no_iex | :refused}
+  def take_over(identifier, opts) do
+    GenServer.call(@name, {:take_over, identifier, opts}, :infinity)
+  end
+
+  ## Callbacks
+
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]: `@impl true`?

File: lib/iex/lib/iex/broker.ex [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  end
+
+  ## Broker API
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, :ok, name: @name)
+  end
+
+  @doc """
+  Register an IEx server in the broker.
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]:
```suggestion
  Registers an IEx server in the broker.
```

File: lib/iex/lib/iex/broker.ex [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  """
+  @spec register(pid) :: :ok
+  def register(pid) do
+    GenServer.call(@name, {:register, pid})
+  end
+
+  @doc """
+  Respond to a broker take over request.
+
+  Note we need to pass the broker pid to support remote shells.
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]:
```suggestion
  The broker's PID is needed to support remote shells.
```

File: lib/iex/lib/iex/broker.ex [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+  def respond(broker_pid, take_ref, true) do
+    GenServer.call(broker_pid, {:accept, take_ref, Process.group_leader()})
+  end
+
+  def respond(broker_pid, take_ref, false) do
+    GenServer.call(broker_pid, {:refuse, take_ref})
+  end
+
+  @doc """
+  Client requests a take over.
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]:
```suggestion
  Client requests a takeover.
```

File: lib/iex/lib/iex/broker.ex [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  All instances, except shell ones, are registered.
+  """
+  @spec register(pid) :: :ok
+  def register(pid) do
+    GenServer.call(@name, {:register, pid})
+  end
+
+  @doc """
+  Respond to a broker take over request.
```
Comment (fertapric) [inline lib/iex/lib/iex/broker.ex]:
```suggestion
  Client responds to a takeover request.
```

I have doubts regarding the usage of "takeover" vs "take over" on this one

File: lib/iex/lib/iex/pry.ex [hunk 10]
```diff
@@ -69,9 +69,9 @@ defmodule IEx.Pry do
       end
 
     # We cannot use colors because IEx may be off
-    case IEx.Server.take_over(request, opts) do
-      :ok ->
-        :ok
+    case IEx.Broker.take_over(request, [evaluator: self()] ++ opts) do
+      {:ok, server, leader} ->
```
Comment (fertapric) [inline lib/iex/lib/iex/pry.ex]: Minor nitpick: `group_leader`?

File: lib/iex/lib/iex/broker.ex (hunk 2 above)
Comment (josevalim) [reply lib/iex/lib/iex/broker.ex]: We need this because we can't pass user as nil later on. So we either need an `if` or pattern matching.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**